File: ProgrammingAssignment1/Formula.cs
// Author: Anjali Sambhaji Tambe
// File: Formula.cs
// Date: 12/11/2024
// Version: Mac , Rider
// Specification file for Formula class

// Purpose: The Formula class converts given input resources into output
//          resources. It uses a proficiency level and probability mechanism
//          to modify output resources.
//          The class allows clients to simulate different outcome according to
//          predefined probabilities and proficiency level. 
using System;
using System.Collections.Generic;

namespace FormulaProject;

//class Invariants :
//Inputs and outputs must not be null.
//Input and output quantities must be non-negative.
//Proficiency level must always be between 0 and maxProficiencyLevel which defined
//Random chance determines the output multiplier for each formula application.
//Once initialized, inputs and outputs cannot be modified by external code.

public enum FormulaOutcome
{
    Fail,    // Represents a failure (0 output)
    Partial, // Represents partial output (e.g., 75% output)
    Normal,   // Represents normal output (100% output)
    Bonus   // Represents bonus output (e.g., 110% or 120% output)
}

public class Formula
{
    private const int MaxProficiencyLevel = 2; // MaxProficiencyLevel is set as 2 
    private readonly Dictionary<string, int> _inputs; // Dictionary to store input resources
    private readonly Dictionary<string, int> _outputs;// Dictionary to store output resources
    private int _proficiencyLevel; // proficiencyLevel can influence the output of applying formula
    private static Random random = new Random(); // create an instance random so it will create random numbers
    private readonly FormulaOutcome? _outcomeOverride;  // optional outcome override to control the outcome
    
    //constructor for the Formula class with the optional outcomeOverride parameter
    //Precondition: input and output quantities must not be null
    //postcondition: the formula object is initialized with validated inputs and outputsÏ
    public Formula(Dictionary<string, int> inputs, Dictionary<string, int> outputs, FormulaOutcome? outcomeOverride=null)
    {
        if (inputs == null || outputs == null)
        {
            throw new ArgumentException("Inputs and outputs can not be null");
        }

        foreach (var quantity in inputs.Values)
        {
            if (quantity < 0)
            {
                throw new ArgumentException(" input Quantities can not be negative");
            }
        }
        
        foreach (var quantity in outputs.Values)
        {
            if (quantity < 0)
            {
                throw new ArgumentException(" output Quantities can not be negative");
            }
        }
    
        _inputs = new Dictionary<string ,int> (inputs); //Assigning the inputs values
        _outputs = new Dictionary<string, int> (outputs);// Assigning the output values
        _proficiencyLevel = 0;   // firstly initializing the proficiency level to 0
        _outcomeOverride = outcomeOverride;   //assign he optional outcome override
    }
    //public method GetInputs() to query inputs
    //Postcondition : return the dictionary of inputs means name and quantity of resource
    public Dictionary<string, int> GetInputs()
    {
        return new Dictionary<string, int> (_inputs);
    }

    //public method GetOutputs() to query outputs
    //Postcondition : return the dictionary of outputs means name and quantity of resource
    public Dictionary<string, int> GetOutputs()
    {
        return new Dictionary<string, int> (_outputs);
    }
    
    //Apply the formula to determine the result based on proficiency level and random chance
    //Postcondition: Returns the transformed output with adjusted quantities based on the formula's chance
    public Dictionary<string, int> Apply()
    {
        Double multiplier;
        
        //use the outcomeOverride if it is set
       if(_outcomeOverride.HasValue) 
       {
            multiplier = GetMultiplier(_outcomeOverride.Value);
       }
       else
       {
           double chance = random.NextDouble();  // generate the number between 0 and 1
           multiplier = GetMultiplier(GetMultiplierBasedOnProficiencyLevel(chance));
       }

       Dictionary<string, int> result = new Dictionary<string, int>();  // newDictionary to hold calculated output 

       foreach (var output in _outputs)
       {
           int newQuantity = (int)Math.Round(output.Value * multiplier); // it will give whole number
           result[output.Key] = newQuantity;
       }
       return result;     //Return final output result
    }

    // Determines the output multiplier based on the proficiency level and chance
    // Precondition: Chance must be between 0 and 2.
    // Postcondition: Returns the multiplier (0, 0.75, 1, or bonus multiplier) based on proficiencyLevel
    private FormulaOutcome GetMultiplierBasedOnProficiencyLevel(double chance)
    {
        if (_proficiencyLevel == 0)
        {
            if (chance < 0.25) return FormulaOutcome.Fail;
            if (chance < 0.50) return FormulaOutcome.Partial;
            if (chance < 0.95) return FormulaOutcome.Normal;
            return FormulaOutcome.Bonus;
        }

        if (_proficiencyLevel == 1)
        {
            if (chance < 0.25) return FormulaOutcome.Fail;
            if(chance < 0.45) return FormulaOutcome.Partial;
            if (chance < 0.95) return FormulaOutcome.Normal;
            return FormulaOutcome.Bonus;
        }
        
        if (_proficiencyLevel == 2)
        {
            if (chance < 0.15) return FormulaOutcome.Fail;
            if(chance < 0.30) return FormulaOutcome.Partial;
            if (chance < 0.85) return FormulaOutcome.Normal;
            return FormulaOutcome.Bonus;
        }

        return FormulaOutcome.Normal;
    }

    // Increases the proficiency level by 1, up to the maximum level
    // Precondition: Proficiency level must not exceed maxProficiencyLevel.
    // Postcondition: Proficiency level is incremented by 1, unless already at max level.
    public void IncreaseProficiencyLevel()
    {
        if (_proficiencyLevel <= MaxProficiencyLevel)
        {
            _proficiencyLevel++;
        }
        else
        {
            throw new InvalidOperationException("proficiency is already at maximum level");
        }
    }

    // Returns the current proficiency level
    // Postcondition: The current proficiency level is returned
    public int GetProficiencyLevel()
    {
        return _proficiencyLevel;
    }
    
    // Returns double mapped to outcome enum
    //precondition: methods excepts the valid formulaOutcome enum value as input
    //postcondition: returns the double alue corresponding o the given formulaOutcome
    private double GetMultiplier(FormulaOutcome outcome)
    {
        switch (outcome)
        {
            case FormulaOutcome.Fail:
                return 0.0;    // Fail = 0 output
            case FormulaOutcome.Partial:
                return 0.75;   // 75% output
            case FormulaOutcome.Normal:
                return 1.0;    // 100% output
            case FormulaOutcome.Bonus:
                return 1.1;    // 110% output
            default:
                return 1.0;    // Default
        }
    }
    
    // Implementation Invariants:
    // - The proficiency level must always remain between 0 and maxProficiencyLevel.
    // - Inputs and outputs remain immutable after object creation.
    // - Random chance is used for output calculations and cannot be manipulated directly by clients.
}

File: Formula.cpp
#include "Formula.h"
#include "cmath"
#include "random"
#include <iostream>

////////////////////////Class Invariants//////////////////
//inputs and Outputs Must Not Be Empty: The inputs and outputs maps must contain at least one valid entry.
// If either is empty, the constructor will throw an exception.
//Input and Output Quantities Must Be Non-Negative: Each quantity in the inputs and outputs maps must be greater than
// or equal to 0. Negative values are not allowed.
//Input and Output Resource Names Must Not Be Empty: The keys in the inputs and outputs maps
// (representing resource names) must not be empty strings.
//Proficiency Level Must Be Between 0 and maxProficiencyLevel: The proficiencyLevel must always be within the valid
// range, from 0 up to maxProficiencyLevel. Any attempt to exceed this range is disallowed.
//Inputs and Outputs Are Immutable After Initialization: Once the Formula object is created, the inputs and outputs
// cannot be modified by external code, preserving the integrity of the object's state.


using namespace std;

// Constructor for formula class
Formula::Formula(const unordered_map<string, int> inputs,
                 const unordered_map<string, int> outputs) {
    if (inputs.empty() || outputs.empty()) {
        throw invalid_argument("inputs and outputs can not be null");
    }
    for (const auto &quantity: inputs) {
        if (quantity.first.empty()) {
            throw invalid_argument("Input resource names can not be empty");
        }
        if (quantity.second < 0) {
            throw invalid_argument("Input quantities can not be neg");
        }
    }

    for (const auto &quantity: outputs) {
        if (quantity.first.empty()) {
            throw invalid_argument("Output resource names can not be empty");
        }
        if (quantity.second < 0) {
            throw invalid_argument("Output quantities can not be neg");
        }
    }

    this->inputs = inputs;
    this->outputs = outputs;
    proficiencyLevel = 0;  // Initial proficiency level set to 0
}

// Get the inputs
// Preconditions: None
// Postconditions: Returns the inputs map
unordered_map<string, int> Formula::getInputs() const {
    return inputs;
}

// Get the outputs
// Preconditions: None
// Postconditions: Returns the outputs map
unordered_map<string, int> Formula::getOutputs() const {
    return outputs;
}

// Apply the formula and adjust based on proficiency
// Preconditions: inputs and outputs must be properly initialized.
// Postconditions: Returns a map of adjusted output resources.
unordered_map<string, int> Formula::apply() const {
    random_device generator;
    double chance = uniform_real_distribution<double>(0.0, 1.0)(generator);
    double multiplier = getMultiplierBasedOnProficiencyLevel(chance);

    unordered_map<string, int> result;

    for (const auto &output: outputs) {
        int newQuantity = static_cast<int>(round(output.second * multiplier));
        result[output.first] = newQuantity;
    }
    return result;
}

// Increase proficiency level
void Formula::increaseProficiencyLevel() {
    if (proficiencyLevel < maxProficiencyLevel) {
        proficiencyLevel++;
    }
}

// Get current proficiency level
int Formula::getProficiencyLevel() {
    return proficiencyLevel;
}

// Private method to determine multiplier based on chance
double Formula::getMultiplierBasedOnProficiencyLevel(double chance) const {
    if (proficiencyLevel == 0) {
        if (chance < 0.25) return 0;
        if (chance < 0.50) return 0.75;
        if (chance < 0.95) return 1.0;
        return 1.1;
    } else if (proficiencyLevel == 1) {
        if (chance < 0.25) return 0;
        if (chance < 0.45) return 0.75;
        if (chance < 0.95) return 1.0;
        return 1.1;
    } else if (proficiencyLevel == 2) {
        if (chance < 0.15) return 0;
        if (chance < 0.30) return 0.75;
        if (chance < 0.85) return 1.0;
        return 1.2;
    }

    return 1.0;  // Default to 1
}

//////////////////Implementation Invariants////////////////////
//Proficiency Level Remains Between 0 and maxProficiencyLevel: The increaseProficiencyLevel method ensures the
// proficiencyLevel does not exceed the allowed maximum.
//Inputs and Outputs Remain Immutable After Object Creation: The inputs and outputs maps are set during construction
// and are not modified thereafter, ensuring the consistency of these values.
//Random Chance Is Used for Output Calculations and Cannot Be Manipulated Externally: The apply method internally
// generates random values, which influence the formula's output adjustments. This randomness is managed within the
// class and cannot be directly controlled by clients

